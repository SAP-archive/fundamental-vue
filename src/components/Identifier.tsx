import { Component, Prop } from 'vue-property-decorator';
import { Api } from '@/api';
import { componentName } from '@/util';
import { Color, Colors, backgroundColorClassName } from '@/lib';
import TsxComponent from '@/vue-tsx';

const sizeMapping = {
    xxs: 'Extra Extra Small',
    xs: 'Extra Small',
    s: 'Small',
    m: 'Medium (default)',
    l: 'Large',
    xl: 'Extra Large',
    xxl: 'Extra Extra Large',
};
type IdentifierSize = keyof (typeof sizeMapping);
const IdentifierSizes = Object.keys(sizeMapping) as IdentifierSize[];

interface Props {
    icon?: string | null;
    url?: string | null;
    size?: IdentifierSize;
    circle?: boolean;
    transparent?: boolean;
    thumbnail?: boolean;
    backgroundColor?: Color | null;
}

@Component({ name: componentName('Identifier') })
@Api.Component('Identifier')
@Api.defaultSlot('Text displayed by the Identifier.')
export class Identifier extends TsxComponent<Props> {
    @Api.Prop('icon name', prop => prop.type(String))
    @Prop({ type: String, default: null, required: false })
    public icon!: string | null;

    @Api.Prop('image url', prop => prop.type(String))
    @Prop({ type: String, default: null, required: false })
    public url!: string | null;

    @Api.Prop('size', prop => prop.type(String).acceptValues(...IdentifierSizes))
    @Prop({ type: String, default: 'm', required: false })
    public size!: IdentifierSize;

    @Api.Prop('is displayed as circle', prop => prop.type(Boolean))
    @Prop({ required: false, default: false, type: Boolean })
    public circle!: boolean;

    @Api.Prop('is displayed without background', prop => prop.type(Boolean))
    @Prop({ required: false, default: false, type: Boolean })
    public transparent!: boolean;

    @Api.Prop('is displayed with background image', prop => prop.type(Boolean))
    @Prop({ required: false, default: false, type: Boolean })
    public thumbnail!: boolean;

    @Api.Prop('background color', prop => prop.type(String).acceptValues(...Colors))
    @Prop({ required: false, default: null, type: String })
    public backgroundColor!: Color | null;

    public render() {
        return (
            <span class={this.classes} role='presentation' style={this.style}>
                {this.$slots.default}
            </span>
        );
    }

    private get style() {
        const url = this.url;
        if (url == null || !this.thumbnail) {
            return {};
        }
        return {
            'background-image': `url(${this.url})`,
        };
    }

    private get classes() {
        const backgroundColorClasses =
            this.backgroundColor == null ? {} : { [backgroundColorClassName(this.backgroundColor)]: true };
        const iconClass = this.icon == null ? {} : { [`sap-icon--${this.icon}`]: true };
        return {
            ...iconClass,
            ...backgroundColorClasses,
            'fd-identifier--xxs': this.size === 'xxs',
            'fd-identifier--xs': this.size === 'xs',
            'fd-identifier--s': this.size === 's',
            'fd-identifier--m': this.size === 'm',
            'fd-identifier--l': this.size === 'l',
            'fd-identifier--xl': this.size === 'xl',
            'fd-identifier--xxl': this.size === 'xxl',
            'fd-identifier--transparent': this.transparent,
            'fd-identifier--circle': this.circle,
            'fd-identifier--thumbnail': this.thumbnail,
        };
    }
}
