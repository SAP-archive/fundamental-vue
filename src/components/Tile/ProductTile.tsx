import { Slot, Component, Prop, Base } from '@/core';

interface Props {
    url?: string | null;
    title?: string | null;
    isButton?: boolean;
}

@Component('ProductTile')
@Slot('content', 'Product Content')
export class ProductTile extends Base<Props> {
    @Prop('image url', { type: String, default: null, required: true })
    public url!: string | null;

    @Prop({ type: String, default: null })
    public title!: string | null;

    @Prop('render product tile as a button', { type: Boolean, default: false })
    public isButton!: boolean;

    @Prop('disable product tile', { type: Boolean, default: false })
    public disabled!: boolean;

    private get style() {
        const url = this.url;
        if (url == null) {
            return {};
        }
        return {
            'background-image': `url(${this.url})`,
        };
    }

    public render() {
        const content = this.$slots.content;
        const title = this.title;
        const isButton = this.isButton;
        const disabled = this.disabled;

        return (
            <div class='fd-product-tile' role={isButton ? 'button' : null} aria-disabled={disabled}>
                <div class='fd-product-tile__media' style={this.style} />
                <div class='fd-product-tile__content'>
                    {title && <h2 class='fd-product-tile__title'>{title}</h2>}
                    {content}
                </div>
            </div>
        );
    }
}
