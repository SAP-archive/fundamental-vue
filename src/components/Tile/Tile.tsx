import { Color, Colors, backgroundColorClassName } from '@/lib';
import { Slot, Component, Prop, Base } from '@/core';

interface Props {
    media?: any | null;
    title?: string | null;
    description?: string | null;
    actions?: any | null;
    isButton?: boolean;
    rowSpan?: number | null;
    colSpan?: number | null;
    backgroundColor?: Color | null;
}

@Component('Tile')
@Slot('media', 'Tile Media')
@Slot('actions', 'Tile Actions')
export class Tile extends Base<Props> {
    @Prop({ type: String, default: null })
    public title!: string | null;

    @Prop({ type: String, default: null })
    public description!: string | null;

    @Prop('row span', { type: Number, default: null })
    public rowSpan!: number | null;

    @Prop('column span', { type: Number, default: null })
    public colSpan!: number | null;

    @Prop('background color', { acceptableValues: Colors, default: null, type: String })
    public backgroundColor!: Color | null;

    @Prop('render tile as a button', { type: Boolean, default: false })
    public isButton!: boolean;

    @Prop('disable tile', { type: Boolean, default: false })
    public disabled!: boolean;

    public render() {
        const media = this.$slots.media;
        const actions = this.$slots.actions;
        const title = this.title;
        const description = this.description;
        const isButton = this.isButton;
        const disabled = this.disabled;

        return (
            <div class={this.classObject} role={isButton ? 'button' : null} aria-disabled={disabled}>
                {media && <div class='fd-tile__media'>{media}</div>}
                <div class='fd-tile__content'>
                    {title && <h2 class='fd-tile__title'>{title}</h2>}
                    {description && <p>{description}</p>}
                </div>
                {actions && <div class='fd-tile__actions'>{actions}</div>}
            </div>
        );
    }

    get classObject() {
        const rowSpanObject = () => {
            const rowSpan = this.rowSpan;
            if (typeof rowSpan === 'number') {
                return { [`fd-has-grid-row-span-${rowSpan}`]: true };
            }
            return {};
        };

        const colSpanObject = () => {
            const colSpan = this.colSpan;
            if (typeof colSpan === 'number') {
                return { [`fd-has-grid-column-span-${colSpan}`]: true };
            }
            return {};
        };

        const backgroundColorClasses =
            this.backgroundColor == null ? {} : { [backgroundColorClassName(this.backgroundColor)]: true };

        return {
            ...rowSpanObject(),
            ...colSpanObject(),
            ...backgroundColorClasses,
            'fd-tile': true,
        };
    }
}
