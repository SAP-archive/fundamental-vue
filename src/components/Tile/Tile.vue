<template>
  <div
    class="fd-tile"
    :class="classes"
    :role="role"
    :aria-disabled="disabled"
    v-fd-background-color="backgroundColor"
  >
    <div v-if="$slots.media" class="fd-tile__media">
      <slot name="media" />
    </div>
    <div class="fd-tile__content">
      <h2 v-if="title != null" class="fd-tile__title">{{ title }}</h2>
      <p v-if="description != null">{{ description }}</p>
    </div>
    <div v-if="$slots.actions" class="fd-tile__actions">
      <slot name="actions" />
    </div>
  </div>
</template>

<script>
import backgroundColor from "./../../directives/background-color";

export default {
  name: "FdTile",
  directives: { "fd-background-color": backgroundColor },
  props: {
    title: { type: String, default: null },
    description: { type: String, default: null },
    rowSpan: { type: Number, default: null },
    colSpan: { type: Number, default: null },
    backgroundColor: { type: String, default: null },
    isButton: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false }
  },
  computed: {
    role() {
      return this.isButton ? "button" : null;
    },
    classes() {
      return [
        this.rowSpan != null ? `fd-has-grid-row-span-${this.rowSpan}` : "",
        this.colSpan != null ? `fd-has-grid-column-span-${this.colSpan}` : ""
      ];
    }
  }
};
</script>
