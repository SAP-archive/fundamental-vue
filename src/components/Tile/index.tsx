export { Tile } from './Tile';
export { ProductTile } from './ProductTile';
