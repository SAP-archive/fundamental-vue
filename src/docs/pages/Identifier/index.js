export default {
  relatedComponents: ["fd-identifier"]
};
