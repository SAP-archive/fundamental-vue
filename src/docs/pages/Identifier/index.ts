import { ExampleCollectionFunction } from '../types';

export const plugin: ExampleCollectionFunction = ({ Identifier }) => {
  return { icon: 'text-formatting', relatedComponents: [Identifier] };
};
