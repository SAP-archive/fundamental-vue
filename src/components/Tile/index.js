import Tile from "./Tile.vue";
import ProductTile from "./ProductTile.vue";
import { pluginify } from "./../../util";
export default pluginify(Tile, ProductTile);
export { Tile, ProductTile };
