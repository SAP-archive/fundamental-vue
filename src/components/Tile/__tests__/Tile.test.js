import { mount } from "@vue/test-utils";
import Tile from "../Tile.vue";
import Identifier from "@/components/Identifier";

describe("Tile", () => {
  const title = "Title";
  const description = "Description";

  const wrapper = mount(Tile, {
    propsData: {
      title,
      description
    }
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders correct props", () => {
    expect(wrapper.find(".fd-tile__title").text()).toEqual(title);
    expect(wrapper.find(".fd-tile__content p").text()).toEqual(description);
  });

  describe("With Background Color", () => {
    const backgroundColor = "accent-2";
    const wrapper = mount(Tile, {
      propsData: {
        title,
        description,
        backgroundColor
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders correct background color", () => {
      expect(
        wrapper.classes(`fd-has-background-color-${backgroundColor}`)
      ).toBe(true);
    });
  });

  describe("As Button", () => {
    const wrapper = mount(Tile, {
      propsData: {
        title,
        description,
        isButton: true
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has role as a button", () => {
      expect(wrapper.attributes("role")).toEqual("button");
    });
  });

  describe("Disabled", () => {
    const wrapper = mount(Tile, {
      propsData: {
        title,
        description,
        disabled: true
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("is disabled", () => {
      expect(wrapper.attributes("aria-disabled")).toEqual("true");
    });
  });

  describe("With Slot Media", () => {
    const wrapper = mount(Tile, {
      propsData: {
        title,
        description
      },
      slots: {
        media: {
          render(h) {
            return h(Identifier);
          }
        }
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has media slot", () => {
      expect(wrapper.contains(".fd-tile__media")).toBe(true);
    });
  });

  describe("Row Col Span", () => {
    const rowSpan = 3;
    const colSpan = 5;

    const wrapper = mount(Tile, {
      propsData: {
        title,
        description,
        rowSpan,
        colSpan
      }
    });

    it("renders correctly", () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders correct row & col span", () => {
      expect(wrapper.contains(`.fd-has-grid-row-span-${rowSpan}`)).toBe(true);
      expect(wrapper.contains(`.fd-has-grid-column-span-${colSpan}`)).toBe(
        true
      );
    });
  });
});
