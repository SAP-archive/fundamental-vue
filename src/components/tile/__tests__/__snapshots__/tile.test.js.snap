// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders correctly as button: button tile 1`] = `
<div
  aria-disabled="false"
  class="fd-tile"
>
  <div
    class="fd-tile__content"
  >
    <h3
      class="fd-tile__title"
    >
      Title
    </h3>
     
    <p>
      Description
    </p>
  </div>
</div>
`;

exports[`Tile renders correctly when disabled: renders disabled correctly 1`] = `
<div
  aria-disabled="false"
  class="fd-tile"
>
  <div
    class="fd-tile__content"
  >
    <h3
      class="fd-tile__title"
    >
      Title
    </h3>
     
    <p>
      Description
    </p>
  </div>
</div>
`;

exports[`Tile renders correctly: default tile 1`] = `
<div
  aria-disabled="false"
  class="fd-tile"
>
  <div
    class="fd-tile__content"
  >
    <h3
      class="fd-tile__title"
    >
      Title
    </h3>
     
    <p>
      Description
    </p>
  </div>
</div>
`;
